src/storage-controller/src/collection_mgmt.rs

```diff
@@ -192,13 +347,476 @@ where
+        self.cmd_rx.close();
+
+        // Get as many waiting senders as possible.
+        'collect: while let Ok((_batch, sender)) = self.cmd_rx.try_recv() {
+            senders.push(sender);
+
+            // Note: because we're shutting down the sending side of `rx` is no
+            // longer accessible, and thus we should no longer receive new
+            // requests. We add this check just as an extra guard.
+            if senders.len() > CHANNEL_CAPACITY {
+                // There's not a correctness issue if we receive new requests,
+                // just unexpected behavior.
+                tracing::error!("Write task channel should not be receiving new requests");
+                break 'collect;
+            }
+        }
+
+        // Notify them that this collection is closed.
+        //
+        // Note: if a task is shutting down, that indicates the source has been
+        // dropped, at which point the identifier is invalid. Returning this
+        // error provides a better user experience.
+        notify_listeners(senders, || Err(StorageError::IdentifierInvalid(self.id)));
+    }
+
+    async fn handle_updates(
+        &mut self,
+        batch: Vec<(
+            Vec<(Row, i64)>,
+            oneshot::Sender<Result<(), StorageError<T>>>,
+        )>,
+    ) -> Result<(), ()> {
+        // To rate limit appends to persist we add artifical latency, and will
+        // finish no sooner than this instant.
+        let batch_duration_ms = match self.id {
+            GlobalId::User(_) => {
+                Duration::from_millis(self.user_batch_duration_ms.load(Ordering::Relaxed))
+            }
+            // For non-user collections, always just use the default.
+            _ => STORAGE_MANAGED_COLLECTIONS_BATCH_DURATION_DEFAULT,
+        };
+        let use_batch_now = Instant::now();
+        let min_time_to_complete = use_batch_now + batch_duration_ms;
+
+        tracing::debug!(
+            ?use_batch_now,
+            ?batch_duration_ms,
+            ?min_time_to_complete,
+            "batch duration",
+        );
+
+        // Reset the interval which is used to periodically bump the uppers
+        // because the uppers will get bumped with the following update. This
+        // makes it such that we will write at most once every `interval`.
+        //
+        // For example, let's say our `DEFAULT_TICK` interval is 10, so at `t +
+        // 10`, `t + 20`, ... we'll bump the uppers. If we receive an update at
+        // `t + 3` we want to shift this window so we bump the uppers at `t +
+        // 13`, `t + 23`, ... which resetting the interval accomplishes.
+        self.upper_tick_interval.reset();
+
+        let (rows, responders): (Vec<_>, Vec<_>) = batch.into_iter().unzip();
+
+        let rows = rows.into_iter().flatten().collect_vec();
+
+        self.to_write.extend(rows.iter().cloned());
+        self.desired.extend(rows.iter().cloned());
+
+        // TODO: Maybe don't do it every time?
```

Comment:
You could consider using the `ConsolidatingVec` that's used by persist_sink. It only performs (amortized) O(log n) consolidation work per inserted update, at the cost of using up to twice the amount of memory of the fully consolidated vector. It's trading off CPU for memory which I'm not sure makes sense in this context.